ftl/code/controllers/subsystem/ftl_navigation.dm

```diff
@@ -0,0 +1,220 @@
+// 	ddd
+
+
+/datum/controller/subsystem/ftl_navigation/proc/get_sector(sectorid)
+	for(var/datum/sector/S in current_sector.connected_sectors)
+		if(S.id == sectorid) //Try sectors we can jump to
+			return S
+		for(var/datum/sector/S2 in S.connected_sectors)
+			if(S2.id == sectorid) //Try those sectors
+				return S2
+	for(var/datum/sector/S3 in visited_sectors)
+		if(S3.id == sectorid) //As a last ditch, try sectors we ever visited.
+			return S3
+	ftl_state = FTL_IDLE
+	WARNING("No sector ID found. Stopped spoolups/jumps just incase")
+
+
+
+
+/datum/controller/subsystem/ftl_navigation/proc/ftl_init_spoolup(sectorid)
+	if(ftl_state != FTL_IDLE) return
+	ftl_state = FTL_SPOOLUP //Setting early to prevent multiple spoolups
+	var/datum/sector/S = get_sector(sectorid)
+	if(!S) return
+	to_sector = S
+	addtimer(CALLBACK(src, .proc/qdel_old_systems, current_sector, to_sector), 20)
+	current_sector = null
+	spool_up() //Begin FTL
+	return TRUE
+
+
+/datum/controller/subsystem/ftl_navigation/proc/qdel_old_systems(var/datum/sector/oldS, var/datum/sector/toS)
+	for(var/datum/sector/S in oldS.connected_sectors) //Cleans up unvisited sectors.
+		oldS.connected_sectors -= S
+		if(S != toS) qdel(S)
+
+
+/datum/controller/subsystem/ftl_navigation/proc/generate_connecting_sectors(var/datum/sector/S)
+	var/datum/sector/N
+	var/new_sector
+	for(var/a in 1 to rand(1,3))
+		new_sector = pickweight(sector_types)
+		N = new new_sector
+		if(N.unique) sector_types -= new_sector //If the sector is unique remove it from the spawn pool
+		S.connected_sectors += N
+	S.generated_sectors = TRUE
+
+
+
+/datum/controller/subsystem/ftl_navigation/proc/ftl_transition(var/ftl_start = TRUE)
+	message_admins("Changing FTL effects")
+	ftl_parallax(ftl_start)
+	ftl_handle_transit_turfs(ftl_start)
+	if(ftl_start)
+		ftl_state = FTL_JUMPING
+		message_admins("FTL has begun")
+		addtimer(CALLBACK(src, .proc/ftl_transition, FALSE), 100)
+	else leave_ftl()
+
+
+/datum/controller/subsystem/ftl_navigation/proc/leave_ftl()
+	current_sector = to_sector
+	current_sector.visited = TRUE
+	visited_sectors += to_sector
+	to_sector = null
+	for(var/datum/sector/S in current_sector.connected_sectors)
+		generate_connecting_sectors(S)
+	ftl_state = FTL_IDLE
+	message_admins("FTL has ended")
+
+
+/datum/controller/subsystem/ftl_navigation/proc/ftl_parallax(var/ftl_start = TRUE)
+	for(var/area/A in ship_areas)
+		A.parallax_movedir = ftl_start ? 4 : 0
+		for(var/atom/movable/AM in A)
+			if(length(AM.client_mobs_in_contents))
+				AM.update_parallax_contents()
+
+
+/datum/controller/subsystem/ftl_navigation/proc/ftl_handle_transit_turfs(var/ftl_start = TRUE)
+	var/throw_dir
+	var/flavor_text
+	if(ftl_start)
+		throw_dir = WEST
+		flavor_text = "<span class='notice'>You feel the ship lurch as it enters FTL.</span>"
+	else
+		throw_dir = EAST
+		flavor_text = "<span class='notice'>You feel the ship lurch as it exits FTL.</span>"
+	for(var/z in SSmapping.levels_by_trait(ZTRAIT_STATION))

```

Comment:
How performant is this action because you're looping through the contents of the entire ma

Reply:
Also why don't you use ship_areas for this

Reply:
I need to call begin_ftl() on every transit turf to change how its displayed for people without parallax (not finished this yet)

Reply:
right, also im guessing planets arent functional? TG has lavaplanet in parallax since I added it but you might want to look into changing that to our own model (probably in a seperate pR)

Reply:
Also re performance, this took 0.2 seconds on Delta station

Reply:
did you use the profiler for that or are you saying it froze 0.2 seconds :^)

Reply:
world.timeofday dif